boa/src/exec/mod.rs

```diff
@@ -739,6 +747,31 @@ impl Interpreter {
         }
     }
 
+    /// The abstract operation ToPropertyKey takes argument argument. It converts argument to a value that can be used as a property key.
+    /// https://tc39.es/ecma262/#sec-topropertykey
+    #[allow(clippy::wrong_self_convention)]
+    pub fn to_property_key(&mut self, value: &Value) -> Value {
+        let key = self.to_primitive(value, Some("string"));
+        if key.is_symbol() {
+            key
+        } else {
+            self.to_string(&key)
+        }
+    }
+
+    /// https://tc39.es/ecma262/#sec-hasproperty
+    pub fn has_property(&self, obj: &Value, key: &Value) -> bool {
+        if let Some(obj) = obj.as_object() {
+            if !Property::is_property_key(key) {
+                false
+            } else {
+                obj.has_property(key)
+            }
+        } else {
+            false
+        }
+    }
+
```

Comment:
We only check for the objects property, we should also check the property chain.

`"toString" in {}` should return `true`.

Reply:
I'm under the assumption that `has_property` should be going up the prototype chain.

I've looked into it, and [it is](https://github.com/jasonwilliams/boa/blob/master/boa/src/builtins/object/internal_methods_trait.rs#L37). It seems like my assumption is broken somewhere down the line.

For this empty object, it makes a call to [get_prototype_of](https://github.com/jasonwilliams/boa/blob/master/boa/src/builtins/object/internal_methods_trait.rs#L181) which calls `get_internal_slot` with the `PROTOTYPE` constant (which is `"prototype"`. There is no `prototype` field in the internal slots of this object, so this example is evaluating to `false`.

However, the base Object is within the `__proto__` internal slot.

I would assume that `get_prototype_of` would return with the base object that all other object inherit from when there are no other prototype.

I'm confused on this point because the standard is a little cryptic to me: as [\[\[GetPrototypeOf\]\]()](https://tc39.es/ecma262/#sec-ordinary-object-internal-methods-and-internal-slots-getprototypeof) just calls [OrdinaryGetPrototypeOf](https://tc39.es/ecma262/#sec-ordinarygetprototypeof) which simply returns `O.[[Prototype]]`.

From looking at the rest of the codebase, it isn't clear to me whether `INSTANCE_PROTOTYPE` (`"__proto__"`) or `PROTOTYPE` should be used when the standard says `[[Prototype]]`.

Reply:
> From looking at the rest of the codebase, it isn't clear to me whether INSTANCE_PROTOTYPE ("__proto__") or PROTOTYPE should be used when the standard says [[Prototype]].

When the standard says [[Prototype]] they mean internal slot, so that’s “__proto__” for us (or PROTOTYPE_INSTANCE). 

“prototype” is a regular property so would not be in double brackets.

Does this help?

Reply:
> For this empty object, it makes a call to get_prototype_of which calls get_internal_slot with the PROTOTYPE constant (which is "prototype". There is no prototype field in the internal slots of this object, so this example is evaluating to false.

Yeah I’m sure this is wrong, it should be checking for \_\_proto\_\_ not prototype

Reply:
> > For this empty object, it makes a call to get_prototype_of which calls get_internal_slot with the PROTOTYPE constant (which is "prototype". There is no prototype field in the internal slots of this object, so this example is evaluating to false.
> 
> Yeah I’m sure this is wrong, it should be checking for __proto__ not prototype

Changing from `PROTOTYPE` to `INSTANCE_PROTOTYPE` now gives the desirable output `true` for `"toString" in {}`

Reply:
I see - thanks a lot you two! I'll add the fix for INSTANCE_PROTOTYPE.